daemons/attrd/attrd_commands.c

```diff
@@ -955,29 +885,141 @@ attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
+        }
+
+        /* If dampening changed, do an immediate write-out,
+         * otherwise repeated dampening changes would prevent write-outs
+         */
+        write_or_elect_attribute(a);
+    }
+
+    return pcmk_rc_ok;
+}
+
+static void
+update_minimum_protocol_ver(const char *value)
+{
+    int ver;
+
+    pcmk__scan_min_int(value, &ver, 0);
+
+    if (ver > 0 && (minimum_protocol_version == -1 || ver < minimum_protocol_version)) {
+        minimum_protocol_version = ver;
+        crm_trace("Set minimum attrd protocol version to %d",
+                  minimum_protocol_version);
+    }
+}
+
+static attribute_t *
+populate_attribute(xmlNode *xml, const char *attr)
+{
+    attribute_t *a = NULL;
+    bool update_both = false;
+
+    const char *op = crm_element_value(xml, PCMK__XA_TASK);
+
+    if (attr == NULL) {
+        crm_warn("Could not update attribute: peer did not specify name");
+        return NULL;
+    }
+
+    // NULL because PCMK__ATTRD_CMD_SYNC_RESPONSE has no PCMK__XA_TASK
+    update_both = pcmk__str_eq(op, PCMK__ATTRD_CMD_UPDATE_BOTH,
+                               pcmk__str_null_matches | pcmk__str_casei);
+
+    // Look up or create attribute entry
+    a = g_hash_table_lookup(attributes, attr);
+    if (a == NULL) {
+        if (update_both || pcmk__str_eq(op, PCMK__ATTRD_CMD_UPDATE, pcmk__str_casei)) {
+            a = create_attribute(xml);
+        } else {
+            crm_warn("Could not update %s: attribute not found", attr);
+            return NULL;
+        }
+    }
+
+    // Update attribute dampening
+    if (update_both || pcmk__str_eq(op, PCMK__ATTRD_CMD_UPDATE_DELAY, pcmk__str_casei)) {
+        int rc = update_attr_dampening(a, xml, attr);
+
+        if (rc != pcmk_rc_ok || !update_both) {
+            return NULL;
+        }
+    }
+
+    return a;
+}
+
+void
+attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
+{
+    attribute_t *a = NULL;
+    const char *attr = crm_element_value(xml, PCMK__XA_ATTR_NAME);
+    const char *value = crm_element_value(xml, PCMK__XA_ATTR_VALUE);
+    gboolean is_force_write = FALSE;
 
-        crm_node_t *known_peer = crm_get_peer(v->nodeid, host);
+    crm_element_value_int(xml, PCMK__XA_ATTR_FORCE, &is_force_write);
+
+    a = populate_attribute(xml, attr);
+    if (attr == NULL) {
+        return;
+    }
 
-        crm_trace("Learned %s has node id %s",
-                  known_peer->uname, known_peer->uuid);
-        if (attrd_election_won()) {
-            write_attributes(FALSE, FALSE);
+    // If no host was specified, update all hosts recursively
+    if (host == NULL) {
+        GHashTableIter vIter;
+
+        crm_debug("Setting %s for all hosts to %s", attr, value);
+        xml_remove_prop(xml, PCMK__XA_ATTR_NODE_ID);
+        g_hash_table_iter_init(&vIter, a->values);
+        while (g_hash_table_iter_next(&vIter, (gpointer *) & host, NULL)) {
+            attrd_peer_update(peer, xml, host, filter);
```

Comment:
It should be possible to call update_attr_on_host() here instead

Reply:
That change happens in 4a4b9a74a1 but I can move it to this patch if you'd prefer.